### DISCUSSION TITLE
What is a proper way to assign a terminal application to a specific workspace?

### DISCUSSION THREAD

**Discussion by someApprentice** [i3/i3#5043] [Q&A] [answered]
I want to assign `ranger` to 3rd workspace, but `alacritty`  by default doesn't define any specific window property for this program.

Here's a print what xprop gave me:

    WM_HINTS(WM_HINTS):
    _NET_WM_DESKTOP(CARDINAL) = 0
    _NET_WM_STATE(ATOM) = _NET_WM_STATE_FOCUSED
    WM_STATE(WM_STATE):
    		window state: Normal
    		icon window: 0x0
    WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, _NET_WM_PING
    _NET_WM_ICON(CARDINAL) = 	Icon (64 x 64):
    	(not shown)
    
    WM_NORMAL_HINTS(WM_SIZE_HINTS):
    		program specified size: 800 by 600
    _NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_NORMAL
    WM_CLIENT_MACHINE(STRING) = "host"
    _NET_WM_PID(CARDINAL) = 213624
    WM_CLASS(STRING) = "Alacritty", "Alacritty"
    XdndAware(ATOM) = BITMAP
    _MOTIF_WM_HINTS(_MOTIF_WM_HINTS) = 0x2, 0x0, 0x1, 0x0, 0x0
    _NET_WM_NAME(UTF8_STRING) = "Alacritty"
    WM_NAME(STRING) = "Alacritty"

I found a way to define the window class using the `--class` argument and `-e` to launch a terminal application with the desired instance, for example `alacritty --class ranger -e ranger`. But writing every time this command would be tedious, and also I want to run this program from `dmenu`. So I want to every time I write `ranger` to dmenu it run alacritty with these arguments.

I found a [solution](https://askubuntu.com/a/1178639) to this problem by replacing the `dmenu_run` script with `i3-dmenu-desktop` and editing the desktop entry, e.g:

    [Desktop Entry]
    Type=Application
    Name=ranger
    Comment=Launches the ranger file manager
    Icon=utilities-terminal
    Terminal=false
    Exec=alacritty --class ranger -e ranger
    Categories=ConsoleOnly;System;FileTools;FileManager
    MimeType=inode/directory;
    Keywords=File;Manager;Browser;Explorer;Launcher;Vi;Vim;Python

So, my question is: I set the `Terminal` key to `false`, because every time I launched ranger from dmenu it also launched an extra terminal with the ranger process, and when I closed it, it also closed ranger itself. Setting this key fixed the issue. But wouldn't changing this key break something? I mean ranger is a terminal application wouldn't be rational to set this key to false.

I also changed the `Name` key in `chromium.desktop` to `Chrome` to launch it from a faster shortcut from dmenu. Is it fine too?

**Comment by Airblader** [comment, marked as answer]
> I found a way to define the window class using the --class argument and -e to launch a terminal application with the desired instance

That is the correct way. The window manager only sees a window, it doesn't know what application is running inside it — it doesn't even know that the window is a terminal.

> But writing every time this command would be tedious

You can also define an alias to not have to write that every time.

> But wouldn't changing this key break something?

No, the change makes perfect sense. Previously the desktop file only launched ranger, but without a supporting graphical terminal that'd be useless. You changed the command to open alacrity, which takes care of opening the window instead.

> Is it fine too?

Sure, it's your file. You can edit desktop entries however you see fit.

**Reply by someApprentice** [reply]
Thank you for your answer! I have been looking for a solution to this problem for a couple of days and haven't found a complete answer in one place. Your answer made it comprehensive. I hope this topic will help others with this problem.

> You can also define an alias to not have to write that every time.

Do you mean an alias in `.bashrc`?

**Reply by Airblader** [reply]
Yes

**Reply by carere** [reply]
Your question helped me 😄

**Reply by someApprentice** [reply]
> Your question helped me 😄

I very glad that it is! But all thanks should go to the author of this wm and the answerer of this question **Airblader**
